File: src/converter_java/converter/Converter.java
package converter;

import java.util.Arrays;

import static java.lang.Math.max;
import static java.lang.Math.pow;

public class Converter {
    public TreeNode convertArrayToTree(int[] array) {
        TreeNode root = new TreeNode(array[0]);
        generateTreeFromArray(array, root, 0);
        return root;
    }

    private void generateTreeFromArray(int[] array, TreeNode root, int i) {
        int n = array.length;
        int lpos = 2 * i + 1, rpos = 2 * i + 2;
        if (lpos < n && array[lpos] != 0) {
            root.left = new TreeNode(array[lpos]);
            generateTreeFromArray(array, root.left, lpos);
        }
        if (rpos < n && array[rpos] != 0) {
            root.right = new TreeNode(array[rpos]);
            generateTreeFromArray(array, root.right, rpos);
        }
    }

    public int[] convertTreeToArray(TreeNode root) {
        int arraySize = getArraySizeForTree(root);
        int[] array = new int[arraySize];
        generateArrayFromTree(root, 0, array);
        return arrayStripTrailing(array);
    }

    private int[] arrayStripTrailing(int[] array) {
        int n = array.length;
        if (array[n - 1] != 0) return array;
        int end = n - 1;
        for (int i = n - 1; i >= 0; i--) {
            if (array[i] == 0 && array[i - 1] != 0) {
                end = i;
                break;
            }
        }
        return Arrays.copyOf(array, end);
    }

    private void generateArrayFromTree(TreeNode root, int i, int[] array) {
        if (root == null || i >= array.length) return;
        array[i] = root.val;
        generateArrayFromTree(root.left, 2 * i + 1, array);
        generateArrayFromTree(root.right, 2 * i + 2, array);
    }

    protected int getArraySizeForTree(TreeNode root) {
        // Could use my own pow algorithm.
        return (int) (pow(2, getTreeDepth(root)) - 1);
    }

    protected int getTreeDepth(TreeNode root) {
        return (root == null) ? 0 : 1 + max(getTreeDepth(root.left), getTreeDepth(root.right));
    }
}

File: src/converter_python/converter.py
from typing import List

from nodes import TreeNode, ListNode


class Converter:
    def convert_array_to_tree(self, array: List[int]) -> TreeNode:
        root: TreeNode = TreeNode(array[0])
        self._generate_tree_from_array(array, root, 0)
        return root

    def _generate_tree_from_array(self, array: List[int], root: TreeNode, i: int) -> None:
        n = len(array)
        lpos, rpos = 2 * i + 1, 2 * i + 2
        if lpos < n and array[lpos]:
            root.left = TreeNode(array[lpos])
            self._generate_tree_from_array(array, root.left, lpos)
        if rpos < n and array[rpos]:
            root.right = TreeNode(array[rpos])
            self._generate_tree_from_array(array, root.right, rpos)

    def convert_tree_to_array(self, root: TreeNode) -> List[int]:
        array: List[int] = [0] * (self._get_array_size_for_tree(root))
        self._generate_array_from_tree(root, 0, array)
        return self.array_rstrip(array)

    def _generate_array_from_tree(self, root: TreeNode, i: int, array: List[int]) -> None:
        # Terminator
        if not root or i >= len(array): return

        # Process
        array[i] = root.val

        # Recursive
        self._generate_array_from_tree(root.left, 2 * i + 1, array)
        self._generate_array_from_tree(root.right, 2 * i + 2, array)

    def _get_array_size_for_tree(self, root: TreeNode) -> int:
        # Todo: could use my own pow
        return pow(2, self._get_tree_depth_dfs(root)) - 1

    def _get_tree_depth_dfs(self, root: TreeNode) -> int:
        if not root: return 0
        return 1 + max(self._get_tree_depth_dfs(root.left), self._get_tree_depth_dfs(root.right))

    def array_rstrip(self, array: List[int]) -> List[int]:
        while not array[(-1)]:
            array.pop(-1)
        return array

    def convert_array_to_linkedlist(self, array: List[int]) -> ListNode:
        '''
        ToDo: support creating cyclic LinkedList

        Use it something like this:
        head: ListNode = Converter().convert_array_to_linkedlist(array, pos)

        To represent a cycle in the given linked list, we use an integer pos which represents
        the position (0-indexed) in the linked list where tail connects to. If pos is -1, then
        there is no cycle in the linked list.

        For more, refers to test141_python.test_solutions.py
        https://leetcode.com/articles/linked-list-cycle/
        '''
        dummy_node: ListNode = ListNode(0)
        head: ListNode = ListNode(array[0])
        dummy_node.next = head

        for i in range(1, len(array)):
            head.next = ListNode(array[i])
            head = head.next

        return dummy_node.next

    def convert_linkedlist_to_array(self, head: ListNode) -> List[int]:
        res: List[int] = []

        while head:
            res.append(head.val)
            head = head.next

        return res
